### ISSUE TITLE
Merge ILogger-V1.cs into ILogger.cs

### ISSUE THREAD

**Issue by snakefoot** [NLog/NLog#3432]
The NLog ILogger-interface is crazy huge. Lets slim it down by only providing generic-version.

**Comment by 304NotModified** [comment]
As far as I remember this is because of performance reasons. This has been proposed years ago.

**Comment by 304NotModified** [comment]
Somewhat related: https://github.com/NLog/NLog/pull/375

**Comment by snakefoot** [comment]
> As far as I remember this is because of performance reasons

Yes in the old days in NLog 1.0 there was nothing called generics. So to avoid the params-array-allocation, then one had methods with all possible combination of datatypes.

**Comment by 304NotModified** [comment]
I think the discussion was around NLog 3 and it wasnot a boxing issue as far I remember, but we need to find that issue/retest the performance impact

**Comment by snakefoot** [comment]
Think the only performance issue I can see is that ILogger interface causes all methods to be virtual.


Would be a good idea if ILogger only contained the params-array method. And none of the other (slimmer interface to mock)


Sent from my Sony Xperia


---- Julian Verdurmen wrote ----


I think the discussion was around NLog 3 and it wasnot a boxing issue as far I remember, but we need to find that issue/retest the performance impact

**Comment by 304NotModified** [comment]
> Would be a good idea if ILogger only contained the params-array method. And none of the other (slimmer interface to mock)

True, but there is another option for that:

1. create a slimmer interfaces as parent 
2. provide a wrapper to implement the large interface (it wraps the smaller interface)

no breaking changes, easy to mock.  There is already a slimmer interface.

**Comment by snakefoot** [comment]
> no breaking changes, easy to mock. There is already a slimmer interface.

Ok. I was just curious about the issue mentioned in #375, that implementing an interface would cause logger-methods to become virtual.

Then it would be better if only the methods that takes params-array was coming from interface and becoming virtual :)

**Comment by 304NotModified** [comment]
virtual methods are indeed more expensive, but implementations has been changed over the years

**Comment by snakefoot** [comment]
But still curious about what performance issue you see that might come from removing ILogger-V1.cs ?

**Comment by 304NotModified** [comment]
can't find the issue, the only thing I could think of:

- logger.log(2)

is less expensive with `log(int)` compared to `log<T>(T t)`

**Comment by snakefoot** [comment]
> is less expensive with log(int) compared to log<T>(T t)

Less expensive at compile-time? not sure I understand.

**Comment by 304NotModified** [comment]
runtime. it's in the JIT code. It's only a guess, so we need a benchmark ;)

**Comment by snakefoot** [comment]
Created PullRequest #4044 where `ILogger` is made obsolete to redirect users to a more simple `ILog`-interface.
